pkg/storage/redis.go

```diff
@@ -0,0 +1,366 @@
+package storage
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	goredislib "github.com/go-redis/redis/v8"
+	"github.com/go-redsync/redsync/v4"
+	"github.com/go-redsync/redsync/v4/redis/goredis/v8"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	PONG               = "PONG"
+	RedisScanBatchSize = 1000
+	RedisMutex         = "redis-mutex"
+)
+
+func init() {
+	err := RegisterStorage(&RedisDB{})
+	if err != nil {
+		panic(err)
+	}
+}
+
+type RedisDB struct {
+	db    *goredislib.Client
+	ctx   context.Context
+	mutex *redsync.Mutex
+}
+
+func (b *RedisDB) Init(i interface{}) error {
+	options := i.(map[string]interface{})
+
+	client := goredislib.NewClient(&goredislib.Options{
+		Addr:     options["address"].(string),
+		Password: options["password"].(string),
+	})
+
+	pool := goredis.NewPool(client)
+	rs := redsync.New(pool)
+
+	b.db = client
+	b.mutex = rs.NewMutex(RedisMutex)
+	b.ctx = context.Background()
+
+	if options["flush"] != nil && options["flush"].(bool) {
+		err := b.db.FlushAll(b.ctx).Err()
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (b *RedisDB) URI() string {
+	return b.db.Options().Addr
+}
+
+func (b *RedisDB) IsOpen() bool {
+	pong, err := b.db.Ping(b.ctx).Result()
+	if err != nil {
+		logrus.Error(err)
+		return false
+	}
+
+	return pong == PONG
+}
+
+func (b *RedisDB) Type() Type {
+	return Redis
+}
+
+func (b *RedisDB) Close() error {
+	return b.db.Close()
+}
+
+func (b *RedisDB) Write(namespace, key string, value []byte) error {
+	nameSpaceKey := getRedisKey(namespace, key)
+
+	if err := b.mutex.Lock(); err != nil {
+		return errors.Wrap(err, "cannot obtain mutex lock")
+	}
+	defer func() {
+		ok, unlockErr := b.mutex.Unlock()
+		if !ok || unlockErr != nil {
+			logrus.Error(unlockErr)
+		}
+	}()
```

Comment:
The `SET` command is an atomic operation in Redis, so locking shouldn't be required for a write operation.
```suggestion
```

Reply:
Ok well in that case we almost don't need a mutex. Does the UpdateValueAndOperation even still need it?